### PR TITLE
[feat] support different spark version

### PR DESCRIPTION
**Pull request by GuoPhilipse**
### What is the purpose of the change

support different spark version in linkis

### Related issues/PRs

Related issues: #4144
Related pr:#4146

### Brief change log

- allow spark version to be specific by user
- update scala version to 2.12 to go with spark3
- remove OOM related error 
- fix duplicated method `foreachPartition (ROW)`
- fix ``SparkSQLHistoryParser`` to fit for different spark version

### Checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [Linkis mailing list](https://linkis.apache.org/community/how-to-subscribe) first)
- [ ] **If this is a code change**: I have written unit tests to fully verify the new behavior.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -113,6 +114,7 @@
     <curator.version>2.7.1</curator.version>
     <guava.version>30.0-jre</guava.version>
     <netty.version>4.1.86.Final</netty.version>
+    <hive-storage-api.version>2.4.0</hive-storage-api.version>
```
Comment (jackxu2011) [inline pom.xml:117]: hive-storage-api.version should be 2.7.2 in spark 3.2?  add this to spark3.2 profile

**Review by jackxu2011 — COMMENTED** [review]
is the hive-storage-api.version should be upgrade in spark-3.2?

File: pom.xml (hunk 1 above)
Comment (GuoPhilipse) [reply pom.xml]: after talk with code auth .three oom error can be removed, if so we don't need to add this dependency.
 @rarexixi  Could you please help double confirm?

**Review by jackxu2011 — APPROVED** [review]
LGTM

**Comment by rarexixi** [comment]
LGTM

**Comment by GuoPhilipse** [comment]
Thanks @rarexixi and @jackxu2011 for your review~

**Comment by XXwhite** [comment]
Hello，Do I need to compile globally? Can I only compile spark engine？

**Comment by XXwhite** [comment]
> > 
> 
> for now we are moving underlying verion in ec module to global, i suggest you may complile with single module by specify `-pl` for example:
> 
> `mvn install package -Pspark-3.2 -Phadoop-3.3 -Dspark.version=3.3.1 -pl linkis-engineconn-plugins/spark`

I have a problem in your way，I tried to solve it but failed：Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.7.1:compile (scala-compile-first) on project linkis-engineplugin-spark: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.7.1:compile failed: Compilation failed

另外我想问一下：我是否可以只将spark引擎升级到3.x版本，其他组件版本不变。

**Comment by GuoPhilipse** [comment]
> 

Could you attach linkis version and compile command?  after linkis 1.3.2(will release soon) it will support different spark version like above,before linkis 1.3.2,you can refer to https://linkis.apache.org/docs/latest/deployment/version-adaptation to do adaptation

**Comment by XXwhite** [comment]
> > 
> 
> Could you attach linkis version and compile command? after linkis 1.3.2(will release soon) it will support different spark version like above,before linkis 1.3.2,you can refer to https://linkis.apache.org/docs/latest/deployment/version-adaptation to do adaptation

No any change，maven version is 3.8.1
cmd:  mvn install package -Pspark-3.2 -Phadoop-3.3 -Dspark.version=3.3.1 -pl linkis-engineconn-plugins/spark

**Comment by GuoPhilipse** [comment]
> 

Can you provide your compiling linkis version or source code branch?

**Comment by XXwhite** [comment]
> > 
> 
> Can you provide your compiling linkis version or source code branch?

1.3.2  I'm using the latest version，clone  master

**Comment by GuoPhilipse** [comment]
> mvn install package -Pspark-3.2 -Phadoop-3.3 -Dspark.version=3.3.1 -pl linkis-engineconn-plugins/spark

the above command is only an example ,you may need to adjust the profile and spark.version if necessary,for your spark version is 3.0.0 and hadoop2.7.2, you can try
 `mvn install package -Pspark-3.2 -Dspark.version=3.0.0 -pl linkis-engineconn-plugins/spark`
image

**Comment by XXwhite** [comment]
> > mvn install package -Pspark-3.2 -Phadoop-3.3 -Dspark.version=3.3.1 -pl linkis-engineconn-plugins/spark
> 
> the above command is only an example ,you may need to adjust the profile and spark.version if necessary,for your spark version is 3.0.0 and hadoop2.7.2, you can try `mvn install package -Pspark-3.2 -Dspark.version=3.0.0 -pl linkis-engineconn-plugins/spark` image

Thank you for your reply, I'll try it

**Pull request merged**